### PR TITLE
feat: invite users from the contact list to the chat 

### LINKED ISSUE #123: feat: invite users from the contact list to the chat
At the right chat settings\contacts panel

### PR DESCRIPTION
**Pull request by evgeniy-ac**
Closes #123

image

### CODE REVIEW

**Review by alexyakunin — COMMENTED** [review]
Left some comments

File: src/dotnet/UI.Blazor/Components/Modal/modal.css [hunk 1]
```diff
@@ -33,3 +33,7 @@
 .custom-modal-class .modal-body {
     @apply md:mb-5;
 }
+
+.blazored-modal {
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/Modal/modal.css:37]: Isn't it the same selector as above one?

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            FilteredChatContacts = AllChatContacts.ToArray();
+        } else {
+            FilteredChatContacts = AllChatContacts
+                .Where(x => x.Author.Name.Contains(filter, StringComparison.OrdinalIgnoreCase))
+                .ToArray();
+        }
+    }
+
+    private void OnContactClick(ContactViewModel contact) {
+        if (contact.Existing) {
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor:126]: We don't use brackets for a single-statement block -- unless you need to put a comment inside as well :)

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    private async Task OnCancel() {
+        await ModalInstance.CancelAsync();
+    }
+
+    private void OnFilter(ChangeEventArgs args) {
+        var filter = args.Value?.ToString();
+        if (string.IsNullOrWhiteSpace(filter)) {
+            FilteredChatContacts = AllChatContacts.ToArray();
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor:117]: No brackets for single-statement blocks

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            .Select(x => x.Author.Id.Value)
+            .ToArray();
+        var createChatAuthorsCommand = new IChatAuthors.CreateChatAuthorsCommand(Session, Chat.Id, userIds);
+        await ChatAuthors.CreateChatAuthors(createChatAuthorsCommand, CancellationToken.None);
+
+        await ModalInstance.CloseAsync();
+    }
+
+    private async Task OnCancel() {
+        await ModalInstance.CancelAsync();
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor:111]: Must be expression bodied member

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        CanInvite = false;
+
+        var userIds = AllChatContacts
+            .Where(x => !x.Existing && x.Checked)
+            .Select(x => x.Author.Id.Value)
+            .ToArray();
+        var createChatAuthorsCommand = new IChatAuthors.CreateChatAuthorsCommand(Session, Chat.Id, userIds);
+        await ChatAuthors.CreateChatAuthors(createChatAuthorsCommand, CancellationToken.None);
+
+        await ModalInstance.CloseAsync();
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor:107]: No ConfigureAwait

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+    private async Task OnInvite() {
+        CanInvite = false;
+
+        var userIds = AllChatContacts
+            .Where(x => !x.Existing && x.Checked)
+            .Select(x => x.Author.Id.Value)
+            .ToArray();
+        var createChatAuthorsCommand = new IChatAuthors.CreateChatAuthorsCommand(Session, Chat.Id, userIds);
+        await ChatAuthors.CreateChatAuthors(createChatAuthorsCommand, CancellationToken.None);
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor:105]: No ConfigureAwait

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    private bool CanInvite { get; set; }
+
+    protected override async Task OnInitializedAsync() {
+        var existingChatUserIds = new HashSet<string>(
+            await ChatAuthors.GetUserIds(Session, Chat.Id, CancellationToken.None),
+            StringComparer.Ordinal);
+
+        var contacts = await UserContacts.GetContacts(Session, CancellationToken.None);
+        var resolveUserAuthors = contacts.Select(x => UserAuthors.Get(x.TargetUserId, true, CancellationToken.None));
+        var userAuthors = await Task.WhenAll(resolveUserAuthors);
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor:73]: No ConfigureAwait

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    private ContactViewModel[] FilteredChatContacts { get; set; } = Array.Empty<ContactViewModel>();
+    private ContactViewModel[] AllChatContacts { get; set; } = Array.Empty<ContactViewModel>();
+    private bool CanInvite { get; set; }
+
+    protected override async Task OnInitializedAsync() {
+        var existingChatUserIds = new HashSet<string>(
+            await ChatAuthors.GetUserIds(Session, Chat.Id, CancellationToken.None),
+            StringComparer.Ordinal);
+
+        var contacts = await UserContacts.GetContacts(Session, CancellationToken.None);
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor:71]: No ConfigureAwait

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    [Parameter] public Chat Chat { get; set; } = null!;
+    [CascadingParameter] BlazoredModalInstance ModalInstance { get; set; } = null!;
+
+    private ContactViewModel[] FilteredChatContacts { get; set; } = Array.Empty<ContactViewModel>();
+    private ContactViewModel[] AllChatContacts { get; set; } = Array.Empty<ContactViewModel>();
+    private bool CanInvite { get; set; }
+
+    protected override async Task OnInitializedAsync() {
+        var existingChatUserIds = new HashSet<string>(
+            await ChatAuthors.GetUserIds(Session, Chat.Id, CancellationToken.None),
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor:68]: No ConfigureAwait

File: src/dotnet/Chat.Service/ChatAuthors.cs [hunk 10]
```diff
@@ -140,6 +150,23 @@ public virtual async Task AddToContacts(IChatAuthors.AddToContactsCommand comman
...
+    public virtual async Task CreateChatAuthors(IChatAuthors.CreateChatAuthorsCommand command, CancellationToken cancellationToken)
+    {
+        if (Computed.IsInvalidating())
+            return; // It just spawns other commands, so nothing to do here
+
+        var chatUserIds = await Backend.GetUserIds(command.ChatId, cancellationToken).ConfigureAwait(false);
+        var existingUserIds = new HashSet<string>(chatUserIds, StringComparer.Ordinal);
+        foreach (var userId in command.UserIds) {
+            if (existingUserIds.Contains(userId)) {
+                continue;
```
Comment (alexyakunin) [inline src/dotnet/Chat.Service/ChatAuthors.cs:162]: We don't add brackets for single-statement block.

**Review by alexyakunin — COMMENTED** [review]
More comments

File: src/dotnet/Chat.Service/ChatAuthors.cs [hunk 11]
```diff
@@ -128,6 +128,16 @@ public virtual async Task<ImmutableArray<string>> GetAuthorIds(Session session,
         return await Backend.GetAuthorIds(chatId, cancellationToken).ConfigureAwait(false);
     }
 
+    // [ComputeMethod]
+    public virtual async Task<ImmutableArray<string>> GetUserIds(Session session, string chatId, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (evgeniy-ac) [reply src/dotnet/Chat.Service/ChatAuthors.cs:132]: To show your contacts that already in a chat

File: src/dotnet/Chat.Service/ChatAuthors.cs [hunk 12]
```diff
@@ -128,6 +128,16 @@ public virtual async Task<ImmutableArray<string>> GetAuthorIds(Session session,
...
+    {
+        var user = await Auth.GetUser(session, cancellationToken).ConfigureAwait(false);
+        if (!user.IsAuthenticated)
+            return ImmutableArray<string>.Empty;
```
[the remark this answers is not in the document]
Comment (evgeniy-ac) [reply src/dotnet/Chat.Service/ChatAuthors.cs:136]: I can throw an exception but there are no exceptions after `if (!user.IsAuthenticated)` in the code

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor (hunk 4 above)
Comment (evgeniy-ac) [reply src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor]: done

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor (hunk 6 above)
Comment (evgeniy-ac) [reply src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor]: done

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor (hunk 7 above)
Comment (evgeniy-ac) [reply src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor]: done

File: src/dotnet/Chat.Service/ChatAuthors.cs (hunk 10 above)
Comment (evgeniy-ac) [reply src/dotnet/Chat.Service/ChatAuthors.cs:162]: done

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor (hunk 5 above)
Comment (evgeniy-ac) [reply src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor]: done

File: src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor (hunk 2 above)
Comment (evgeniy-ac) [reply src/dotnet/Chat.UI.Blazor/Components/InviteChatAuthor.razor]: done

File: src/dotnet/Chat.Service/ChatAuthors.cs [hunk 13]
```diff
@@ -140,6 +150,23 @@ public virtual async Task AddToContacts(IChatAuthors.AddToContactsCommand comman
         _ = await UserContactsBackend.GetOrCreate(user2Id, userId1, cancellationToken).ConfigureAwait(false);
     }
 
+    public virtual async Task CreateChatAuthors(IChatAuthors.CreateChatAuthorsCommand command, CancellationToken cancellationToken)
+    {
+        if (Computed.IsInvalidating())
+            return; // It just spawns other commands, so nothing to do here
+
+        var chatUserIds = await Backend.GetUserIds(command.ChatId, cancellationToken).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (evgeniy-ac) [reply src/dotnet/Chat.Service/ChatAuthors.cs:158]: It's deduplication. I want to exclude users that already in a chat.

**Comment by evgeniy-ac** [comment]
image

**Review by alexis-kochetov — APPROVED** [review]

**Comment by evgeniy-ac** [comment]
image

**Pull request merged**